### PR TITLE
Changes to builders for v7 store feature

### PR DESCRIPTION
**Pull request by tmeasday**
Part of #15871

Telescoping on https://github.com/storybookjs/storybook/pull/16004

## What I did

- Added a new entry point for builders in `storyStoreV7` mode. 
- The new entry point creates a `WebPreview`, and passes it:
  - A `fetchStoryIndex` function (basic fetch for now)
  - A `getProjectAnnotations` function that composes the various `config` entries (i.e. `preview.js` and addons)
  - An `importFn` that is an `import()` with some extra webpack trickery.
- It also deals with HMR in a (hopefully) simple way.

### CODE REVIEW

File: lib/builder-webpack4/src/preview/virtualModuleModernEntry.js.handlebars [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+if (module.hot) {
+  module.hot.accept('./{{storiesFilename}}', () => {
+    // importFn has changed so we need to patch the new one in
+    preview.onImportFnChanged({ importFn });
+  });
+
+  module.hot.accept([{{#each configs}}'{{this}}',{{/each}}], () => {
+    // getProjectAnnotations has changed so we need to patch the new one in
+    preview.onGetProjectAnnotationsChanged({ getProjectAnnotations });
```
Comment (shilman) [inline lib/builder-webpack4/src/preview/virtualModuleModernEntry.js.handlebars:42]: great 💯

File: lib/core-common/src/utils/to-importFn.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+export function toImportFnPart(entry: NormalizedStoriesEntry) {
+  const { path: base, match } = toRequireContext(entry.glob);
+
+  const webpackIncludeRegex = new RegExp(match.substring(1));
+
+  // NOTE: `base` looks like './src' but `path`, (and what micromatch expects)
+  // is something that starts with `src/`. So to strip off base from path, we
+  // need to drop `base.length - 1` chars.
+  return dedent`
+      async (path) => {
```
Comment (shilman) [inline lib/core-common/src/utils/to-importFn.ts:15]: extra spaces 😉

File: lib/core-server/src/utils/stories-json.ts [hunk 3]
```diff
@@ -70,7 +69,9 @@ export async function useStoriesJson(router: any, options: Options) {
   });
   const globs = stories.map((s) => s.glob);
   extractStoriesJson(storiesJson, globs, options.configDir);
+  console.log('extracted');
```
Comment (shilman) [inline lib/core-server/src/utils/stories-json.ts:72]: console.log

File: lib/core-server/src/utils/stories-json.ts [hunk 4]
```diff
@@ -70,7 +69,9 @@ export async function useStoriesJson(router: any, options: Options) {
...
   router.use('/stories.json', async (_req: any, res: any) => {
+    console.log('getting stories.json');
```
Comment (shilman) [inline lib/core-server/src/utils/stories-json.ts:74]: console.log

**Review by shilman — COMMENTED** [review]
Good one! This seems much more sane than the previous version

File: lib/core-server/src/utils/stories-json.ts (hunk 3 above)
Comment (tmeasday) [reply lib/core-server/src/utils/stories-json.ts:72]:
```suggestion
```

File: lib/core-server/src/utils/stories-json.ts (hunk 4 above)
Comment (tmeasday) [reply lib/core-server/src/utils/stories-json.ts:74]:
```suggestion
```

**Comment by shilman** [comment]
Merged as part of https://github.com/storybookjs/storybook/pull/15871

**Pull request closed without merging**